OrbitService/ProducerSideServiceImpl.cpp

```diff
@@ -0,0 +1,329 @@
+
+  std::atomic<bool> receive_events_thread_exited = false;
+
+  // This thread is responsible for writing on stream, and specifically for
+  // sending StartCaptureCommands and StopCaptureCommands to the connected producer.
+  std::thread send_commands_thread{&ProducerSideServiceImpl::SendCommandsThread,
+                                   this,
+                                   context,
+                                   stream,
+                                   &all_events_sent_received,
+                                   &receive_events_thread_exited};
+
+  // This thread is responsible for reading from stream, and specifically for
+  // receiving CaptureEvents and AllEventsSent messages.
+  std::thread receive_events_thread{&ProducerSideServiceImpl::ReceiveEventsThread, this, context,
+                                    stream, &all_events_sent_received};
+  receive_events_thread.join();
+
+  // When receive_events_thread exits because stream->Read(&request) fails,
+  // it means that the producer has disconnected: ask send_commands_thread to exit, too.
+  receive_events_thread_exited = true;
+  send_commands_thread.join();
+
+  {
+    absl::MutexLock lock{&server_contexts_mutex_};
+    server_contexts_.erase(context);
+  }
+
+  LOG("Finished handling ReceiveCommandsAndSendEvents for a CaptureEventProducer");
+  return grpc::Status::OK;
+}
+
+void ProducerSideServiceImpl::SendCommandsThread(
+    grpc::ServerContext* context,
+    grpc::ServerReaderWriter<orbit_grpc_protos::ReceiveCommandsAndSendEventsResponse,
+                             orbit_grpc_protos::ReceiveCommandsAndSendEventsRequest>* stream,
+    bool* all_events_sent_received, std::atomic<bool>* receive_events_thread_exited) {
+  // As a result of initializing prev_capture_status to kCaptureFinished,
+  // an initial StartCaptureCommand is sent
+  // if service_state_.capture_status is actually CaptureStatus::kCaptureStarted,
+  // and an initial StopCaptureCommand is sent (with little effect)
+  // if service_state_.capture_status is actually CaptureStatus::kCaptureStopping.
+  CaptureStatus prev_capture_status = CaptureStatus::kCaptureFinished;
+
+  // This loop keeps track of changes to service_state_.capture_status using
+  // conditional critical sections on service_state_mutex_ and updating prev_capture_status,
+  // and sends StartCaptureCommands and StopCaptureCommands accordingly.
+  // It exits when one of *receive_events_thread_exited and service_state_.exit_requested is true,
+  // or when Write fails (because the producer disconnected or because the context was cancelled).
+  while (true) {
+    // This is set when ReceiveEventsThread has exited. At that point this thread should also exit.
+    if (*receive_events_thread_exited) {
+      return;
+    }
+
+    service_state_mutex_.Lock();
+    if (service_state_.exit_requested) {
+      service_state_mutex_.Unlock();
+      return;
+    }
+
+    if (service_state_.capture_status != prev_capture_status) {
```

Comment:
if you reverse this if - you can scope service_state_mutex_ better. I think.

Reply:
Done, it doesn't make the biggest difference but I guess it's a bit better.

Let me know if there's something else that needs to be addressed.